modules/tinymce/src/core/test/ts/browser/html/DomParserTest.ts

```diff
@@ -386,6 +386,52 @@ describe('browser.tinymce.core.html.DomParserTest', () => {
     assert.equal(results.href.nodes[1].attr('href'), '2.gif', 'Parser filter result node(1) attr');
   });
 
+  it('TINY-8888: mutating addNodeFilter -> addAttributeFilter', () => {
+    const parser = DomParser({});
+    parser.addNodeFilter('img', (nodes) => {
+      Arr.each(nodes, (node) => node.attr('src', null));
+    });
+    parser.addAttributeFilter('src', () => {
+      assert.fail('second src filter should not run, because src was removed');
+    });
+    parser.parse('<b>a<img src="1.gif" />b</b>');
+  });
+
+  it('TINY-8888: mutating addNodeFilter -> addNodeFilter', () => {
+    const parser = DomParser({});
+    parser.addNodeFilter('img', (nodes) => {
+      Arr.each(nodes, (node) => node.remove());
+    });
+    parser.addNodeFilter('img', () => {
+      assert.fail('second img filter should not run, because img was removed');
+    });
+    parser.parse('<b>a<img src="1.gif" />b</b>');
+  });
+
+  it('TINY-8888: mutating addAttributeFilter -> addAttributeFilter', () => {
+    const parser = DomParser({});
+    parser.addAttributeFilter('src', (nodes) => {
+      Arr.each(nodes, (node) => node.attr('src', null));
+    });
+    parser.addAttributeFilter('src', () => {
+      assert.fail('second src filter should not run, because src was removed');
+    });
+    parser.parse('<b>a<img src="1.gif" />b</b>');
+  });
+
+  it('TINY-8888: mutating addAttributeFilter only removes matching nodes', () => {
+    const parser = DomParser({});
+    parser.addAttributeFilter('src', (nodes) => {
+      nodes[0].attr('src', null);
+    });
+    let ranIdFilter = false;
+    parser.addAttributeFilter('src', () => {
+      ranIdFilter = true;
+    });
+    parser.parse('<b>a<img src="1.gif" />b<img src="1.gif" />c</b>');
```

Comment:
Wondering if it would be good to verify the correct node still exists. 
```suggestion
    parser.addAttributeFilter('src', (nodes) => {
      ranIdFilter = true;
      assert.lengthOf(nodes, 1);
      assert.equal(node[0].attr('src'), '2.gif');
    });
    parser.parse('<b>a<img src="1.gif" />b<img src="2.gif" />c</b>');
```

Reply:
yeah I guess that makes sense. I'll merge the suggestion then double check it fails correctly.